### PR TITLE
scripts: always check certificates

### PR DESCRIPTION
**Pull request by roysjosh**
Signed-off-by: Joshua Roys <roysjosh@gmail.com>

### CODE REVIEW

**Comment by aparcar** [comment]
Please elaborate why you want this change. An attack doesn't seem feasible as long as you don't break sha256.

**Comment by aparcar** [comment]
Apart from that, please always add a commit message describing what and why your doing things.

**Comment by roysjosh** [comment]
I'd call it the principle of least surprise. Or defense in depth. There are [other attack vectors](https://curl.se/docs/CVE-2009-0037.html) this would protect against.

**Review by oldium — APPROVED** [review]

File: scripts/download.pl [hunk 1]
```diff
@@ -82,8 +82,8 @@ ($)
 	}
 
 	return $have_curl
-		? (qw(curl -f --connect-timeout 20 --retry 5 --location --insecure), shellwords($ENV{CURL_OPTIONS} || ''), $url)
```
Comment (svlobanov) [inline scripts/download.pl:85]: I think it should be configurable (with checking certs by default). sometimes certs are not updated by ops/scripts and you may want to download without cert check and rely on checksums

**Comment by Neustradamus** [comment]
To follow

**Comment by jow-** [comment]
The recent fallout from the Letsencrypt root certificate expiration has shown that forced SSL verification like that can cause unforeseen troubles several years down the road. I'm against this change. We authenticate downloads by checksum, TLS identity does not matter.

**Comment by roysjosh** [comment]
Respectfully, I disagree. Yes, root CAs need upkeep. That requires future work. But it took me only a minute of searching to find an attack vector that this could protect against. If we have a switch that can disable encryption, users can work around future issues like the LetsEncrypt cutover. As a counterpoint, I would argue that it is likely that within a decade we'll have to switch off of sha256. Security is a moving target.

**Comment by roysjosh** [comment]
In my former life I was a sysadmin & I did occasionally forget to renew/rotate certs (sorry). I get that you're aiming for a good balance of security & availability for the project & I respect that. I'm just very passionate about reducing attack surfaces as that is part of my current day job. I hope you would at least consider a switch that can be toggled on/off. Start with it off but have a few contributors try turning it on for a month or two & we go from there.

**Comment by Ansuel** [comment]
Why not make this an optional feature? And make it configurable in menuconfig?

**Comment by aparcar** [comment]
@jow- I'm in favor of this patch. Our own mirrors are always the last once to try and in case of another Let's Encrypt like fallout, we can adopt our mirrors to fix things.

**Comment by Ansuel** [comment]
@roysjosh can you change this and create an option? Since we have mixed feeling about this i would ask you to set it off by default to handle breakage case.

We can discuss later to turn it on by default but this totally needs to be configurable as we need to do tests sometime :D

**Review by BKPepe — APPROVED** [review]

**Comment by roysjosh** [comment]
```
+ curl -f --connect-timeout 20 --retry 5 --location --insecure https://codeload.github.com/ninja-build/ninja/tar.gz/v1.11.0?/ninja-1.11.0.tar.gz
```

Then set config option:
```
$ diff .config.old .config | grep CERT
> CONFIG_CERTIFICATE_CHECK=y
```

Followed by:
```
+ curl -f --connect-timeout 20 --retry 5 --location  https://codeload.github.com/ninja-build/ninja/tar.gz/v1.11.0?/ninja-1.11.0.tar.gz
```

File: config/Config-build.in [hunk 2]
```diff
@@ -58,6 +58,10 @@ menu "Global build settings"
 		bool "Enable signature checking in opkg"
 		default SIGNED_PACKAGES
 
+	config CERTIFICATE_CHECK
```
Comment (Ansuel) [inline config/Config-build.in:61]: Mhhh maybe DOWNLOAD_CHECK_CERTIFICATE

File: config/Config-build.in [hunk 3]
```diff
@@ -58,6 +58,10 @@ menu "Global build settings"
...
+		bool "Enable TLS certificate verification during package download"
+		default n
```
Comment (BKPepe) [inline config/Config-build.in:63]: Should be enabled by default as discussed in this PR

Comment (roysjosh) [reply config/Config-build.in:63]: Is that the consensus? I've seen both opinions voiced.

Comment (BKPepe) [reply config/Config-build.in:63]: I am afraid that the option to enable it in the future will take some time or somebody will forget to turn it on. Besides that, who will turn this on by default manually? Would he/she know that there is a new option that can be used? From my side, it is better to have this on in the master branch, so we can catch those failures and fix them. Rather be safe and secure than regret, and if someone regrets, he/she can disable it.

**Comment by Ansuel** [comment]
Merged in master with minor addition to the commit description with 90c6e3aedf167b0ae1baf376e7800a631681e69a. Thanks!

**Pull request closed without merging**